extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
 import {Arrow} from './arrow';
 import {Scroller} from './scroller';
 import {toChildArray, useRef, useState} from '../../../src/preact';
+import {useMountEffect} from '../../../src/preact/utils';
 
 /**
- * @param {!JsonObject} props
+ * @param {!BaseCarouselDef.Props} props
  * @return {PreactDef.Renderable}
  */
 export function BaseCarousel(props) {
   const {
-    'arrowPrev': arrowPrev,
-    'arrowNext': arrowNext,
-    'children': children,
-    'defaultSlide': defaultSlide,
-    'loop': loop,
+    arrowPrev,
+    arrowNext,
+    children,
+    defaultSlide,
+    loop,
+    slide,
+    onSlideChange,
+    onMount_,
+    setAdvance,
     ...rest
   } = props;
   const childrenArray = toChildArray(children);
   const {length} = childrenArray;
   const [curSlide, setCurSlide] = useState(defaultSlide || 0);
-  const ignoreProgrammaticScroll = useRef(true);
-  const setRestingIndex = (i) => {
-    ignoreProgrammaticScroll.current = true;
-    setCurSlide(i);
-  };
-  const scrollRef = useRef(null);
+  const slideState = slide ?? curSlide;
+  const ignoreProgrammaticScrollRef = useRef(true);
   const advance = (dir) => {
     const container = scrollRef.current;
     // Modify scrollLeft is preferred to `setCurSlide` to enable smooth scroll.
     // Note: `setCurSlide` will still be called on debounce by scroll handler.
     container./* OK */ scrollLeft += container./* OK */ offsetWidth * dir;
   };
+  useMountEffect(() => {
+    if (onMount_) {
+      onMount_();
+    }
+    if (setAdvance) {
+      setAdvance(advance);
+    }
+  });
+
+  const setRestingIndex = (i) => {
+    ignoreProgrammaticScrollRef.current = true;
```

Comment:
It looks like `ignoreProgrammaticScrollRef` is only used here and inside the `Scroller`. And the `setRestingIndex` is also only used inside the `Scroller`. I think that means that you can safely move `ignoreProgrammaticScrollRef` to the `Scroller`, which would reduce amount of state shared.

Reply:
Done.